lib/internal/modules/esm/hooks.js

```diff
@@ -87,45 +86,40 @@ let importMetaInitializer;
 // [2] `validate...()`s throw the wrong error
 
 class Hooks {
-  #chains = {
-    /**
-     * Prior to ESM loading. These are called once before any modules are started.
-     * @private
-     * @property {KeyedHook[]} globalPreload Last-in-first-out list of preload hooks.
-     */
-    globalPreload: [],
-
-    /**
-     * Phase 1 of 2 in ESM loading.
-     * The output of the `resolve` chain of hooks is passed into the `load` chain of hooks.
-     * @private
-     * @property {KeyedHook[]} resolve Last-in-first-out collection of resolve hooks.
-     */
-    resolve: [
-      {
-        fn: defaultResolve,
-        url: 'node:internal/modules/esm/resolve',
-      },
-    ],
-
-    /**
-     * Phase 2 of 2 in ESM loading.
-     * @private
-     * @property {KeyedHook[]} load Last-in-first-out collection of loader hooks.
-     */
-    load: [
-      {
-        fn: require('internal/modules/esm/load').defaultLoad,
-        url: 'node:internal/modules/esm/load',
-      },
-    ],
-  };
+  #loaderInstances = [];
+  #chains = {};
 
   // Cache URLs we've already validated to avoid repeated validation
   #validatedUrls = new SafeSet();
 
   allowImportMetaResolve = false;
 
+  constructor() {
+    const defaultLoader = 'internal/modules/esm/default_loader';
+    this.addCustomLoader(`node:${defaultLoader}`, require(defaultLoader));
+  }
+
+  #rebuildChain(name) {
+    const chain = this.#chains[name] = [];
+    let i = 0;
+    for (const instance of this.#loaderInstances) {
+      if (typeof instance[name] !== 'function') {
+        continue;
+      }
+      chain.push({
+        loader: instance,
+        fn: instance[name],
+        next: chain[i++ - 1],
+      });
+    }
+  }
+
+  #rebuildChains() {
```

Comment:
This is a performance loss. I think it would be better for there to be only `rebuildChains` and for it to accept a list of chains to rebuild, like

```js
#rebuildChains(...names) {
  if (!names.length) {
    names = new Array('initialize','load','resolve'); // [1]
  }

  const shouldRebuild = { __proto__: null };

  for (let n = names.length - 1; n > -1; n--) {
    const name = names[n];
    shouldRebuild[name]: true;
    this.#chains[name].length = 0; // [2] Might need a primordial?
  }

  for (
    let i = 0, l = this.#loaderInstances.length - 1;
    i < l;
    i++
  ) {
    const {
      initialize,
      load,
      resolve,
    } = this.#loaderInstances[i];

    if (shouldRebuild.initialize) { /* … */ }
    if (shouldRebuild.load) { /* … */ }
    if (shouldRebuild.resolve) { /* … */ }
  }
}
```

```js
this.#rebuildChains('resolve'); // Rebuild only `resolve` chain
this.#rebuildChains('load','resolve'); // Rebuild `load` and `resolve`
this.#rebuildChains(); // Rebuild all chains
```

[1] I believe this (`new Array(…)` instead of `[…]`) will invoke an optimisation in V8 to choose the most appropriate specialised C++ array from the start and also to pre-allocation only exactly the space for these 3 items. If it doesn't, no need for the otherwise uncommon use of `new Array()` instead of `[]`.

[2] We don't need to re-create the array, only empty it. For that, setting its length to 0 is far better.

Reply:
I think we can make this happen 😄 Probably a bit of a µ-optimization considering how often `#rebuildChains` is called and the size of `N`… but it's not a problem to make it a little bit faster. If the actual chain execution was affected I would be more concerned, considering that is called on every `import`/`require`.

Reply:
Mm, it likely won't break the performance bank. But it's easy to avoid the de-op, so why not 🙂